cpp_exercise/src/lib/linuxipclib.cc

```diff
@@ -0,0 +1,262 @@
+     switch (protocolOptions(_protocol))
+    {
+    case ipcprt::Protocol::pipe:
+        ipcpipe(_file_name);
+        break;
+    case ipcprt::Protocol::msgqueue:
+        ipcmsgqueue(_file_name);
+        break;
+    default:
+        throw IPCException("IPCReceiver ERROR:UNKNOWN IPC PROTOCOL");
+        break;
+    }
+
+}
+    
+/**
+ * @brief Command Option Class
+ * 
+ */
+
+ 
+ CommandOption::~CommandOption(){
+
+     _option = 0;
+     optind = 0;
+ }
+
+
+/**
+ * @brief Construct a new Command Option for Handling the terminal arguments
+ * 
+ * 
+ * 
+ * 
+ */
+
+CommandOption::CommandOption(const std::string & program, int argc, char *argv[]): _program(program), _argc(argc){
+    
+     
+   optind = 0;
+   
+
+    
+
+    while(1){
+      
+      _longindex = 0;
+      option longopts[]= {
+			{"help", no_argument, NULL, 'h'},
+			{"msgqueue", no_argument, NULL, 'q'},
+			{"pipe", no_argument, NULL, 'p'},
+			{"shm", required_argument, NULL, 's'},
+			{"file", required_argument, NULL, 'f'},
+			{ 0, 0, 0, 0 }
+      };
+
+      _option = getopt_long(_argc, argv, "h:q:p:s:f:",longopts, &_longindex);
+
+      if(_option == -1)
+      break;
+
+       switch (_option)
+           {
+              case 'h':
+              printHelp();
+              exit(EXIT_SUCCESS);
+                break; 
+              case 'p':
+               _output = "pipe";
+               break;  
+              case 'q':
+              _output = "msgqueue";
+               break; 
+              case 's':
+              throw IPCException("unrecognized command. please use \"--help\" for guide.\n");
+                break;
+               case 'f':
+                _filename = optarg;
+                if ((_filename.size() < 1) || (_filename.size() >40) )
+                {
+                  throw IPCException("unrecognized command. the file name should be between 1 and 30 letters");
+                }
+
+                 break;
+              case ':':
+                throw IPCException("unrecognized command. please use \"--help\" for guide.\n");
+                break;
+                case '?':
+	              throw IPCException("unrecognized command. please use \"--help\" for guide.\n");
+                break;
+                default:
+	              throw IPCException("unrecognized command. please use \"--help\" for guide.\n");
+                break;
+              }
+    }
+
+	}
+
+/**
+ * @brief Construct a new Command Option for Handling the terminal arguments
+ * 
+ * 
+ * 
+ * 
+ */
+  
+/**
+ * @brief Returning the ipc protocol and filename
+ * 
+ * 
+ */
+std::vector<std::string> CommandOption::getCommand(){
+
+  if (_filename.size() == 0 || _output.size() == 0 )
+  {
+    throw IPCException("unrecognized command. filename and protocol should be determined. please use \"--help\" for guide.\n");
+  }
+  
+  _options.push_back(_filename);
```

Comment:
It's pretty not expected from function named 'getCommand' to have side effects. Consider what will happen if user calls it many times

Reply:
I moved the initialization of _options to the constructor and it getCOmmand it will only return the vector